controllers/configmap_controller.go

```diff
@@ -334,3 +334,48 @@ func (r *ConfigMapReconciler) createServicesConfigMap(ctx context.Context, names
 	r.log.Info("Created", "ConfigMap", namespacedName)
 	return windowsServices, nil
 }
+
+// EnsureServicesConfigMapExists ensures that the ServicesConfigMap is present and valid on the cluster during operator
+// bootup. ConfigMapReconciler.createServicesConfigMap() cannot be called in its stead as the cache has not been
+// populated yet, which is why the typed client is used here as it calls the API server directly.
+func (r *ConfigMapReconciler) EnsureServicesConfigMapExists() error {
+	windowsServices, err := r.k8sclientset.CoreV1().ConfigMaps(r.watchNamespace).Get(context.TODO(), servicescm.Name,
+		meta.GetOptions{})
+	if err != nil {
+		if !k8sapierrors.IsNotFound(err) {
+			return err
+		}
+		// If ConfigMap is not found, create it and return
+		windowsServices, err = servicescm.Generate(servicescm.Name, r.watchNamespace)
+		if err != nil {
+			return err
+		}
+		if _, err = r.k8sclientset.CoreV1().ConfigMaps(r.watchNamespace).Create(context.TODO(), windowsServices,
+			meta.CreateOptions{}); err != nil {
+			return err
+		}
+		r.log.Info("Created", "ConfigMap", kubeTypes.NamespacedName{Namespace: r.watchNamespace, Name: servicescm.Name})
```

Comment:
This is repeated code. Please modularize. Or even better you can set a flag `createConfigMap` here and below which indicates the ConfigMap needs to be created; then only create the ConfigMap if the flag is set.

Reply:
Explored both options. Modularizing lead to a much cleaner flow (since we only want to Parse() if we were able to successfully retrieve the configmap).